net/listener_with_tls.go

```diff
@@ -0,0 +1,154 @@
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"net/http"
+
+	"github.com/drand/drand/protobuf/drand"
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/nikkolasg/slog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// NewGRPCListenerForPublicAndProtocolWithTLS creates a new listener for the Public and Protocol APIs over GRPC with TLS.
+func NewGRPCListenerForPublicAndProtocolWithTLS(ctx context.Context, bindingAddr string, certPath, keyPath string, s Service, opts ...grpc.ServerOption) (Listener, error) {
+	lis, err := net.Listen("tcp", bindingAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	x509KeyPair, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	grpcCreds, err := credentials.NewServerTLSFromFile(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, grpc.Creds(grpcCreds))
+	opts = append(opts, grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor))
+	opts = append(opts, grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor))
+	serverOpts := append(opts, grpc.Creds(grpcCreds))
+	grpcServer := grpc.NewServer(serverOpts...)
+	drand.RegisterPublicServer(grpcServer, s)
+	drand.RegisterProtocolServer(grpcServer, s)
+
+	httpServer := buildTLSServer(grpcServer, x509KeyPair)
+	g := &tlsListener{
+		Service:    s,
+		httpServer: httpServer,
+		grpcServer: grpcServer,
+		l:          tls.NewListener(lis, httpServer.TLSConfig),
+	}
+	grpc_prometheus.Register(g.grpcServer)
+	return g, nil
+}
+
+// NewRESTListenerForPublicWithTLS creates a new listener for the Public API over REST with TLS.
+func NewRESTListenerForPublicWithTLS(ctx context.Context, bindingAddr string, certPath, keyPath string, s Service, opts ...grpc.ServerOption) (Listener, error) {
+	lis, err := net.Listen("tcp", bindingAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	x509KeyPair, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	grpcCreds, err := credentials.NewServerTLSFromFile(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, grpc.Creds(grpcCreds))
+	opts = append(opts, grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor))
+	opts = append(opts, grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor))
+	serverOpts := append(opts, grpc.Creds(grpcCreds))
+	grpcServer := grpc.NewServer(serverOpts...)
+	drand.RegisterPublicServer(grpcServer, s)
+
+	o := runtime.WithMarshalerOption("*", defaultJSONMarshaller)
+	gwMux := runtime.NewServeMux(o)
+	proxy := &drandProxy{s}
+	err = drand.RegisterPublicHandlerClient(context.Background(), gwMux, proxy)
+	if err != nil {
+		return nil, err
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/", gwMux)
+
+	httpServer := buildTLSServer(grpcHandlerFunc(grpcServer, mux), x509KeyPair)
+	g := &tlsListener{
+		Service:    s,
+		httpServer: httpServer,
+		grpcServer: grpcServer,
```

Comment:
this 'rest' server is still muxing and talking GRPC as well as HTTP. can it be a simpler "just http" server?